### PR TITLE
[GH-#59] Mnemonic encryption password now submitted by user

### LINKED ISSUE #59: Fix the hardcoded password for encrypting the mnemonic file
When the function for enrypting  the file is called a password should be passed as a parameter and used to encrypt/decript the file holding the mnemonic. The default value should be deleted.

PS. The user password when generating a wallet is it's own password, which is still not obligatory.

### PR DESCRIPTION
**Pull request by Boychev-Dimitar**
Resolves #59 

Added a verification if the password entered is correct. 
Made small code refactoring. 

Current default password is " ", but could be changed to something like "aeternity".

### CODE REVIEW

File: lib/elixir_wallet/elixir_wallet.ex [hunk 1]
```diff
@@ -32,25 +32,31 @@ defmodule Wallet do
   If the wallet was not password protected, just pass the mnemonic_phrase
   """
   @spec import_wallet(String.t(), String.t()) :: String.t()
-  def import_wallet(mnemonic_phrase, password \\ "") do
+  def import_wallet(mnemonic_phrase, password \\ " ") do
```
Comment (gspasov) [inline lib/elixir_wallet/elixir_wallet.ex]: The default value should be "", not " "

File: lib/elixir_wallet/elixir_wallet.ex [hunk 2]
```diff
@@ -32,25 +32,31 @@ defmodule Wallet do
...
   Decrypts a file and displays it's mnemonic phrase 
+  Will only return a phrase if the password is correct  
   ## Examples
       iex> Wallet.load_wallet_file("wallet--2017-10-31-14-54-39", "password")
 	  {:ok,
 	  "spirit\r beach\r smile\r turn\r glance\r whale\r rack\r reflect\r marble\r cover\r enter\r pigeon"}
   """
   @spec load_wallet_file(String.t(), String.t()) :: Tuple.t()
-  def load_wallet_file(file_path, password) do
+  def load_wallet_file(file_path, password \\ " ") do
```
Comment (gspasov) [inline lib/elixir_wallet/elixir_wallet.ex]: Here as well, default should be ""

File: lib/elixir_wallet/elixir_wallet.ex [hunk 3]
```diff
@@ -70,35 +76,46 @@ defmodule Wallet do
...
   @spec get_public_key(String.t(), String.t()) :: String.t()
-  def get_public_key(file_path, password) do
-   {_, mnemonic} = load_wallet_file(file_path, password)
+  def get_public_key(file_path, password \\ " ") do
+  {validation, mnemonic} = load_wallet_file(file_path, password)
      
-   KeyPair.generate_root_seed(mnemonic)
-   |> elem(1)
+  if (validation != :error) do 
+   public_key = Kernel.elem(KeyPair.generate_root_seed(mnemonic), 1)
```
Comment (gspasov) [inline lib/elixir_wallet/elixir_wallet.ex]: Here you should pass the password as second parameter to the KeyPair.generate_root_index func. But the password of the wallet, not the one used to encrypt the file

File: lib/elixir_wallet/elixir_wallet.ex [hunk 4]
```diff
@@ -17,7 +17,7 @@ defmodule Wallet do
       sand trip update spring
   """
   @spec create_wallet(String.t()) :: String.t()
-  def create_wallet(password \\ "") do
+  def create_wallet(password \\ " ") do
 
     mnemonic_phrase = Mnemonic.generate_phrase(GenerateIndexes.generate_indexes)
     save_wallet_file(mnemonic_phrase, password)
```
Comment (gspasov) [inline lib/elixir_wallet/elixir_wallet.ex]: You should use a different password for encrypting the mnemonic phrase, not the one used for creating the wallet.

File: lib/elixir_wallet/elixir_wallet.ex [hunk 5]
```diff
@@ -70,35 +76,46 @@ defmodule Wallet do
...
+  Will only return an address if the password is correct  
   ## Examples
       iex> Wallet.get_address("wallet--2017-10-31-14-54-39", "password")
 	  "1NM51tw1MixFCe64g6ExhCEXnowEGrQ2DE"
   """
   @spec get_address(String.t(), String.t()) :: String.t()
-  def get_address(file_path, password) do
-   get_public_key(file_path, password)
-   |> KeyPair.generate_wallet_address() 
+  def get_address(file_path, password \\ " ") do
```
Comment (gspasov) [inline lib/elixir_wallet/elixir_wallet.ex]: Here as well the default should be "", not " "

File: lib/elixir_wallet/elixir_wallet.ex [hunk 6]
```diff
@@ -70,35 +76,46 @@ defmodule Wallet do
...
   Gets the public key
+  Will only return a public key if the password is correct  
   ## Examples
       iex> Wallet.get_public_key("wallet--2017-10-31-14-54-39", "password")
 	   "04187ACD56365459C0948953DD0D36A93E57A4C78C6E04E1E9DB10EA2C89FF2701DCC1AFB2C30F09ACE265EA11DAFE7E40591182F4EC3E7BB535305551D2A374CB"
   """
   @spec get_public_key(String.t(), String.t()) :: String.t()
-  def get_public_key(file_path, password) do
-   {_, mnemonic} = load_wallet_file(file_path, password)
+  def get_public_key(file_path, password \\ " ") do
```
Comment (gspasov) [inline lib/elixir_wallet/elixir_wallet.ex]: Here as well the default should be "", not " "

File: lib/elixir_wallet/elixir_wallet.ex [hunk 7]
```diff
@@ -32,30 +32,35 @@ defmodule Wallet do
...
-	   {:ok, mnemonic}
-        else
-          Logger.info("Invalid password")
-          {:error, ""}
+          if String.contains?(mnemonic, salt) do
+             mnemonic = String.replace(mnemonic, " " <> salt, "")          
+             {:ok, mnemonic}
+          else
+             Logger.info("Invalid salt")
+             {:error, ""}
```
Comment (thepiwo) [inline lib/elixir_wallet/elixir_wallet.ex]: can you give the correct error messages back and log to error

File: lib/elixir_wallet/elixir_wallet.ex [hunk 8]
```diff
@@ -32,30 +32,35 @@ defmodule Wallet do
...
+          end
+        else 
+           Logger.info("Invalid password")
+           {:error, ""}
```
Comment (thepiwo) [inline lib/elixir_wallet/elixir_wallet.ex]: can you give the correct error messages back and log to error

File: lib/elixir_wallet/elixir_wallet.ex [hunk 9]
```diff
@@ -32,30 +32,35 @@ defmodule Wallet do
...
+          else
+             Logger.info("Invalid salt")
+             {:error, ""}
+          end
+        else 
+           Logger.info("Invalid password")
+           {:error, ""}          
         end  
       {:error, :enoent} ->
         {:error, "The file does not exist."}
```
Comment (thepiwo) [inline lib/elixir_wallet/elixir_wallet.ex]: log the file path as well, that does not exist

File: lib/elixir_wallet/elixir_wallet.ex [hunk 10]
```diff
@@ -71,51 +76,73 @@ defmodule Wallet do
...
      
-  if (validation != :error) do 
-   public_key = Kernel.elem(KeyPair.generate_root_seed(mnemonic), 1) 
-   {:ok, public_key}
-  else
-   {:error, ""}
-  end  
- end
+    if (validation != :error) do 
+      {_, public_key} = Kernel.elem(KeyPair.generate_root_seed(mnemonic, salt), 1) |> Base.decode16()
```
Comment (thepiwo) [inline lib/elixir_wallet/elixir_wallet.ex]: in `generate_master_keys` return keys without base encoding, only encode if needed

Comment (thepiwo) [inline lib/elixir_wallet/elixir_wallet.ex]: pipe to elem(1) if needed, not with Kernel

File: lib/elixir_wallet/elixir_wallet.ex [hunk 11]
```diff
@@ -71,51 +76,73 @@ defmodule Wallet do
...
+  """
+  @spec get_address(String.t(), String.t()) :: Tuple.t()
+  def get_address(file_path, password, salt \\ "") do
+    {validation, public_key} = get_public_key(file_path, password, salt)
+
+    if (validation != :error) do 
+      address = KeyPair.generate_wallet_address(public_key) 
+      {:ok, address}
+    else
+      {:error, ""}
```
Comment (thepiwo) [inline lib/elixir_wallet/elixir_wallet.ex]: can you give the correct error messages back

File: lib/elixir_wallet/elixir_wallet.ex [hunk 12]
```diff
@@ -71,51 +76,73 @@ defmodule Wallet do
...
-   {validation, public_key} = get_public_key(file_path, password)
+  @spec get_private_key(String.t(), String.t()) :: Tuple.t()
+  def get_private_key(file_path, password, salt \\ "") do
+    {validation, mnemonic} = load_wallet_file(file_path, password, salt)
+     
+    if (validation != :error) do 
+      {_, private_key} = Kernel.elem(KeyPair.generate_root_seed(mnemonic, salt), 0) |> Base.decode16()
+      {:ok, private_key}
+    else
+      {:error, ""}
```
Comment (thepiwo) [inline lib/elixir_wallet/elixir_wallet.ex]: can you give the correct error messages back

**Comment by thepiwo** [comment]
please name your commits what they actually do, not "Added requested changes"

File: lib/elixir_wallet/elixir_wallet.ex [hunk 13]
```diff
@@ -65,40 +87,77 @@ defmodule Wallet do
...
+  ## Examples
+      iex> Wallet.get_private_key("wallet--2017-10-31-14-54-39", "password")
+      {:ok, <<100, 208, 92, 132, 43, 104, 6, 55, 125, 18, 18, 215, 98, 8, 245, 12, 78, 92,
+      89, 115, 59, 231, 28, 142, 137, 119, 62, 19, 102, 238, 171, 185>>}
+  """
+  @spec get_private_key(String.t(), String.t()) :: Tuple.t()
+  def get_private_key(file_path, password, pass_phrase \\ "") do
+    {validation, mnemonic} = load_wallet_file(file_path, password, pass_phrase)
+     
+    if (validation != :error) do
```
Comment (thepiwo) [inline lib/elixir_wallet/elixir_wallet.ex]: you should work with `case` here for semantic naming reasons. 
if `load_wallet_file` returns an error the second argument in the tuple will not be a `mnemonic`, but an error message

File: lib/elixir_wallet/elixir_wallet.ex [hunk 14]
```diff
@@ -65,40 +87,77 @@ defmodule Wallet do
...
-  def get_public_key(file_path, password) do
-   {_, mnemonic} = load_wallet_file(file_path, password)
+  @spec get_public_key(String.t(), String.t()) :: Tuple.t()
+  def get_public_key(file_path, password, pass_phrase \\ "") do
+    {validation, mnemonic} = load_wallet_file(file_path, password, pass_phrase)
      
-   KeyPair.generate_root_seed(mnemonic)
-   |> elem(1)
-  end  
+    if (validation != :error) do
```
Comment (thepiwo) [inline lib/elixir_wallet/elixir_wallet.ex]: you should work with case here for semantic naming reasons.
if load_wallet_file returns an error the second argument in the tuple will not be a mnemonic, but an error message

**Comment by thepiwo** [comment]
we will squash and merge this, once its done

**Comment by thepiwo** [comment]
ok, sorry can we do it different, no it is semantically incorrect for both cases. We should name variables what they actually do. `reply` is a bad name in the most cases, as any function call replies something.

This is what i mean, you can pattern match the cases and give the variables seperate names:
```
case load_wallet_file(file_path, password, pass_phrase) do
  {:ok, mnemonic_phrase} ->
    public_key = KeyPair.generate_root_seed(mnemonic_phrase, pass_phrase) |> elem(1)
    {:ok, public_key}
  {:error, message} ->
    {:error, message}
end  
```

**Comment by thepiwo** [comment]
@gspasov can you approve now?

**Comment by thepiwo** [comment]
@meivantodorov @andonov can you a approve and merge?

**Pull request merged**